File: app/src/main/java/com/example/composingapp/views/viewtools/accidentals/FlatLeaf.java
package com.example.composingapp.views.viewtools.accidentals;

import android.graphics.Canvas;
import android.graphics.Paint;
import android.graphics.RectF;

import com.example.composingapp.views.viewtools.positiondict.NotePositionDict;
import com.example.composingapp.views.viewtools.ViewConstants;

import static com.example.composingapp.views.viewtools.ViewConstants.STEM_WIDTH;

public class FlatLeaf extends AccidentalLeaf {
    private Float baseRectBottomY;
    private Float halfSpace;
    private Float baseRectRightX;
    private RectF arcBoundingRect;
    private RectF baseBoundingRect;
    private Paint tempPaint;

    public FlatLeaf(NotePositionDict notePositionDict, Paint paint) {
        super(notePositionDict, paint);
        halfSpace = notePositionDict.getSingleSpaceHeight() / 2;
        initFields();
    }

    private void initFields() {
        mBoundingRectWidth = (float) (W_TO_H_RATIO * mBoundingRectHeight);
        this.centerX = mNotePositionDict.getNoteX() -
                2 * (ViewConstants.NOTE_W_TO_H_RATIO * mNotePositionDict.getNoteVerticalRadius());


        tempPaint = new Paint(mPaint);
        tempPaint.setStrokeWidth(STEM_WIDTH);
        tempPaint.setStyle(Paint.Style.STROKE);
        // Copy the paint colours and overwrite the width/ style types

        // Make the bounding box for the curved and 45 degree stroke
        Float baseRectTopY = centerY - halfSpace;
        baseRectBottomY = centerY + halfSpace;
        baseRectRightX = centerX + mBoundingRectWidth / 2;
        baseBoundingRect = new RectF(centerX, baseRectTopY,
                baseRectRightX, baseRectBottomY);

        // Make the bounding box for the circle (arc stroke)
        arcBoundingRect = new RectF(
                centerX, centerY - halfSpace / 2, baseRectRightX, centerY + halfSpace / 2
        );
    }

    @Override
    public void draw(Canvas canvas) {
        // Reset the color, if needed
        tempPaint.setColor(mPaint.getColor());
        // Draw the vertical line
        canvas.drawLine(centerX, centerY - mBoundingRectHeight / 2, centerX,
                baseRectBottomY, mPaint);
        // Draw the demi-heart
//        canvas.drawRect(baseBoundingRect, tempPaint);
        canvas.drawLine(centerX, baseRectBottomY, baseRectRightX, centerY, tempPaint);
        canvas.drawArc(arcBoundingRect, 0, -180, false, tempPaint);
    }
}

File: app/src/main/java/com/example/composingapp/views/accidentals/FlatLeaf.kt
package com.example.composingapp.views.accidentals

import android.graphics.Canvas
import android.graphics.Paint
import com.example.composingapp.utils.interfaces.PositionDict
import com.example.composingapp.utils.interfaces.componentdrawer.LeafDrawer
import com.example.composingapp.views.accidentals.AccidentalLeaf.W_TO_H_RATIO
import com.example.composingapp.views.viewtools.ViewConstants
import com.example.composingapp.views.viewtools.positiondict.NotePositionDict

object FlatLeaf : LeafDrawer {
    override fun draw(canvas: Canvas, positionDict: PositionDict, paint: Paint) {
        if (positionDict is NotePositionDict) {
            canvas.apply {
                val halfSpace = positionDict.scorePositionDict.singleSpaceHeight / 2
                val mBoundingRectHeight = halfSpace * 4
                val boundingRectWidth = (W_TO_H_RATIO * mBoundingRectHeight).toFloat()
                val centerX = positionDict.noteX -
                        2 * (ViewConstants.NOTE_W_TO_H_RATIO * positionDict.noteVerticalRadius)
                val centerY = positionDict.noteY
                val baseRectBottomY = centerY + halfSpace
                val baseRectRightX = centerX + boundingRectWidth / 2

                // Draw the vertical line
                drawLine(centerX, centerY - mBoundingRectHeight / 2, centerX,
                        baseRectBottomY, paint)

                // Save the old style
                val oldStyle = paint.style
                paint.style = Paint.Style.STROKE
                // Draw the demi-heart
                drawLine(centerX, baseRectBottomY, baseRectRightX, centerY, paint)
                drawArc(centerX, centerY - halfSpace / 2, baseRectRightX,
                        centerY + halfSpace / 2, 0f, -180f, false, paint)
                paint.style = oldStyle
            }
        }
    }
}